feat: initial draft for Proof-of-Stake in Witnet (#96)
witnet/WIPs: draft-aesedepece-pos.md

ARS presence. This rate has been estimated to be around 5%, i.e. once every 20 epochs, the consolidated block will
potentially come from an identity that is completely new to the system, provided that there are enough newcomers trying
to join the network.

#### Witnessing

Witnessing is the ability of participating in resolving a particular oracle query. The algorithm for computing
eligibility for witnessing is known as Reputation-based Proof of Eligibility ([RepPoE]).

In RepPoE, as its name suggests, the reputation score of an identity plays a key role in determining its eligibility.
In its primitive form as introduced in the [2017 Witnet whitepaper][whitepaper], it makes witnessing eligibility
directly proportional to reputation score. However, a number of [other factors][WIP-0016] have been introduced over time
in an attempt to mitigate attack vectors and abuse.

A complete description of the current RepPoE implementation can be found in [WIP-0016]. In regards to this proposal, and
for the sake of simplicity, we can assume that RepPoE is perfectly lineal:

```rust

> // `rf` is the replication factor, i.e. how many witnesses is the requester targetting

Reviewer: ```suggestion
// `rf` is the replication factor, i.e. how many witnesses is the requester targeting
```

into a single VPS, or by running custom versions of the node software that calculate eligibilities for thousands of
identities while only keeping a single instance of the chain state.

This multi-identity or _sock puppet_ behavior goes clearly against the original design goals of the RandPoE algorithm.
But its most serious negative effect is that it especially harms the profitability of more casual node operators who
only run 1 or 2 nodes.

It is not confirmed whether this multi-identity behavior has an impact on witnessing, provided that computing the
witnessing eligibility for each combination of identity and oracle query can prove to be very costly, and that
witnessing rewards are negligible when compared to mining rewards. Moreover, the Sybil resistance properties
provided by collateralization of witnessing actually work against this being feasible or profitable at scale.

#### Late blocks are dangerous

Another phenomenon that has been observed in the Witnet network to a greater or lesser degree is the what is often
referred as _late blocks_. These late blocks are block candidates with extraordinarily low VRF values that are broadcast

> to the network just before an epoch is finishing.

Reviewer: ```suggestion
to the network just before an epoch is finished.
```

  the query's own replication factor (witnesses count), and the commitment round.
- When proposing a block, all stake of the proposing node is used at once. That is, the average age of its stake is
  reset to zero.
- When resolving an oracle query, only the required collateral is used. That is, the average age of the stake is moved
  forward in time proportionally to how that collateral amount compares to the amount of stake that remains unused.
- Block mining and witnessing rewards no longer exist as UTXOs. The rewarded coins are automatically added to the staked
  balance of the block proposer.
- The superblock voting committees are evenly sampled from the `StakesTracker`.
- At any time, stake can be withdrawn from a node, either partially or all at once, through an `UnstakeTransaction`.
- The `UnstakingTransaction` acts as a pre-announcement for the unstaking itself, i.e. the transaction only puts the
  node in an exiting queue, but it still needs to keep voting for superblocks if required to do so, and the unstaked 
  coins are still unspendable and subject to slashing for 2 weeks (_unstaking delay period_).

> - After the 2-weeks unstaking delay period is over, the unstaked coins exist as an UTXO that becomes spendable by the

Reviewer: ```suggestion
- After the 2-weeks unstaking delay period is over, the unstaked coins exist as a UTXO that becomes spendable by the
```

>   staker, and the obligation to kee voting for superblocks ceases.

Reviewer: ```suggestion
  staker, and the obligation to keep voting for superblocks ceases.
```

### Relative timelocks

**()** Timelock values found in the `time_lock` field found in `ValueTransferOutput` MUST be processed in such a way
that values lesser or equal than the UNIX timestamp of the _checkpoint zero_ are treated as relative. That is, those
outputs affected by relative timelocks become spendable only after the amount of seconds specified as `time_lock` has
passed after the timestamp of the block in which the output is created.

In mainnet, being the _checkpoint zero_ equal to `1_602_666_000`, the longest that an output can be time locked is for
`1_602_666_000` seconds, which equates to approximately 50 years, 10 months and 14 days.

> **()** Conversely, `time_lock` values greater than the same _checkpoint zero_ MUST still be trated as absolute. That is,

Reviewer: ```suggestion
**()** Conversely, `time_lock` values greater than the same _checkpoint zero_ MUST still be treated as absolute. That is,
```

transaction's `body`.

**()** For a `StakeTransaction` to be valid, the `authorization` field MUST be a valid ECDSA Secp256k1 signature such
that the public key as recovered from the signature will be treated as the address of the only node that will be allowed
to use the staked coins, aka the _operator_, and the signed message is the only address (namely, 20-bytes
`PublicKeyHash`) that will be allowed to unstake coins from the operator, aka the _withdrawer_.

This `authorization` signature prevents unknown 3rd parties from delegating coins into someone else's node without their
consent.

**()** For a `StakeTransaction` to be valid, its output MUST satisfy one of these conditions:
1. the operator address is not present in the `StakesTracker`, or
2. the specific combination of operator address and withdrawer address has appeared in a previous `StakeTransaction`.

This requirement ensures that each operator address is only nominating 1 withdrawer address. On the contrary, the same
withdrawer address can be used by multiple operators addresses.

**()** A single `ValueTransferOutput` MUST be allowed in a `StakeTransaction`, for the sake of allowing change in case

> that the amount of coins to stake is smalled than the value of the UTXOs consumed in `inputs`.

Reviewer: ```suggestion
that the amount of coins to stake is smaller than the value of the UTXOs consumed in `inputs`.
```

i.e. how long they have been at stake and without being used.

The usage of this metric instead of directly assigning stakers an eligibility that is purely proportional to the number
of staked coins creates a non-negligible opportunity for small stakers to eventually mine or witness, provided that
their stake accrues enough age. At the same time, it deters very big stakers from completely monopolizing the mining
and witnessing, inasmuch it reduces the pace at which they can reuse their big stake.

Eligibilities are thus generally computed as the power of a staker, divided by the global power of the network (the 
aggregate of the power of every other staker), multiplied by some replication factor that allows for redundancy.

**()** The power of an address with 0 stake MUST be exactly 0.

**()** The power of an address with a non-zero stake MUST be computed as the smallest of:
1. the product of its total staked coins and the average age of that stake, as registered in its own entry in 
   `StakesTracker`, and
2. a `MAXIMUM_POWER` constant, initially set to 1/125th of the maximum value of an unsigned 

>    64-but integer number, i.e. `1_475_739_525_896_764_412`.

Reviewer 2: ```suggestion
   64-bit integer number, i.e. `147_573_952_589_676_416`.
```

**()** The global power of the network MUST be computed as the smallest of:
1. the product of the global stake (the total amount of coins that are staked in the network), and the global 
   average age of those stakes, as registered in the `global` entry in `StakesTracker`, and

> 2. the maximum value of an unsigned 64-bit number, i.e. `184_467_440_737_095_551_615`.

Reviewer 2: ```suggestion
2. the maximum value of an unsigned 64-bit number, i.e. `18_446_744_073_709_551_615`.
```

   of them in big-endian format,
2. the power of the block proposer is in the `rf / stakers`th quantile among the powers of all the stakers, and
3. the big-endian value of the VRF output is less than `max_rounds * own_power / (max_power * (rf - max_rounds) - rf * 
   threshold)`,

Where:
- `rf` is the replication factor,
- `stakers` is the number of stakers in the `StakesTracker`,
- `max_rounds` is the maximum number of commitment rounds allowed by the protocol (currently `4`),
- `own_power` is the power of the committer,
- `max_power` is the power of the most powerful staker, and
- `threshold` is the power of the least powerful staker within the `rf/stakers`th quantile from requirement 2 above.

**()** May a node receive and successfully validate a number of concurrent commitments with distinct transaction hashes
for the same data request and epoch greater than the value of the `witnesses` field in the data request, it SHOULD 
prioritize for inclusion in an eventual block candidate those commitments whose VRF output have the lowest big-endian 
value.

**()** Upon consolidating a block candidate, for each of the commitment transactions contained therein, the amount of

> coins in each committer's stake entry in the `StakesTracker` MUST be decremented by an amount equivalent to the 

Reviewer 3: This statement does not align with the latest simulation results I shared and on which this description is based. You have to reset all coins of the selected stakers to age 1. Otherwise the most powerful stakers wull always be eligibile to solve a data request.

  queries.
- Staking is performed through a `StakeTransaction`, which consumes UTXOs and adds the staked coins to the
  `StakesTracker`. I.e. those coins cease to exist as UTXOs.
- Staked coins cannot be moved until they are unstaked.
- Staked coins accrue age for every epoch they have been staked and unused, separately for different protocol 
  capabilities (e.g. mining vs. witnessing).
- Power is the product of total stake of a miner, and the average age of that stake, for different protocol 
  capabilities.
- A node's mining eligibility is roughly proportional to its mining power over the total mining network power, 
  multiplied by a replication factor.
- Similarly, a node's witnessing eligibility is roughly proportional to its witnessing power over the total network 
  witnessing power, multiplied by the query's own replication factor (witnesses count), and the commitment round.
- When proposing a block, all the mining power of the proposing node is used at once. That is, if the block ends up 
  being consolidated, the average "mining age" of its proposer's stake is reset to zero.

> - When resolving an oracle query, only the required collateral is used. That is, the average "witnessing age" of the 

Reviewer 3: See this point: https://github.com/witnet/WIPs/pull/96#discussion_r1316242521

2. the specific combination of operator address and withdrawer address has appeared in a previous `StakeTransaction`.

This requirement ensures that each operator address is only nominating 1 withdrawer address. On the contrary, the same
withdrawer address can be used by multiple operators addresses.

**()** A single `ValueTransferOutput` MUST be allowed in a `StakeTransaction`, for the sake of allowing change in case
that the amount of coins to stake is smaller than the value of the UTXOs consumed in `inputs`.

**()** Only 1 `StakeTransaction` with the same `operator` MAY be allowed into a block.

**()** The `value` field in the `StakeOutput` in `StakeTransaction` MUST be equal or greater than a new
`MINIMUM_STAKE_NANOWITS` constant, initially set to `10_000_000_000_000`, i.e. 10,000 WIT coins.

**()** The weight of a `StakeTransaction` MUST be computed in weight units as `N*133+M*36+105`, where `N` is the number
of `inputs`, and `M` is 0 or 1 depending on whether a `change` output is used.

As a result, the minimum weight for a `StakeTransaction` (one with a single input and no `change` output) is 238 weight
units.

#### UnstakeTransaction

> **()** A new type of transaction, `StakeTransaction`, MUST be introduced, following this schema:

Reviewer 4: `UnstakeTransaction`

**()** Upon consolidating a block candidate, the amount of coins in the proposer's entry in the `StakesTracker` MUST be 
incremented by an amount equivalent to the block reward, plus the addition of all unclaimed value from the transactions 
contained in the block, i.e. the miner fees.

**()** Upon consolidating a block candidate, the average age of the proposer's witnessing entry in the `StakesTracker` 
MUST be reset to zero, i.e. the average epoch MUST be set to the same epoch as the one for which the block was 
consolidated.

#### Witnessing eligibility

**()** For any given data request transaction in the commitment stage, a replication factor `rf` MUST be computed as
the `witnesses` field in the data request multiplied by 2 to the power of the current witnessing round:

`rf = witnesses * 2 ^ round`

**()** For a commitment transaction to be considered for inclusion in a block, it MUST hold an amount of witnessing
power and provide a valid VRF proof that MUST satisfy all the following requirements:
1. the VRF input is the concatenation of the previous block hash, the data request hash, and the current epoch, all 
   of them in big-endian format,

> 2. the witnessing power of the block proposer is in the `rf / stakers`th quantile among the witnessing powers of all 

Author: ```suggestion
2. the witnessing power of the committer is in the `rf / stakers`th quantile among the witnessing powers of all 
```
